README.md
# sparkl

Graph various quadratic surfaces.

## Installation



## Usage

Coming soon.

## Options



## Examples

TBD

## License

Copyright © 2018 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-08-08
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-08-08
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/sparkl/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/sparkl/compare/0.1.0...0.1.1


test/sparkl/core_test.clj
(ns sparkl.core-test
  (:require [clojure.test :refer :all]
            [sparkl.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/sparkl/quadric.clj
(ns sparkl.quadric
  (:require [quil.core :as q]
            [sparkl.surfaces :as s]
            [sparkl.styling :as lnf]))

(defn zero [& args]
  0)

;; options
(def framerate 60)
(def speed 2)
(def rotation (* 2 (Math/PI)))
(def counter (atom 0))
(def frame-count 600)
(def axis? false)
(def render-frames false)

;; surface settings
(def config (:cone s/settings))
(def originX (first (:origin config)))
(def originY (second (:origin config)))
(def xyz-length 128)
(def sheet-size 300)
(def grid-x (:grid-x config))
(def grid-y (:grid-y config))

;; Angles
(def Ax (atom (Math/toRadians (nth (:angles config) 0))))
(def Ay (atom (Math/toRadians (nth (:angles config) 1))))
(def Az (atom (Math/toRadians (nth (:angles config) 2))))
(def orient (atom (Math/toRadians 10)))

(defn copy-sign [val provider]
  "Return val with sign of provider"
  (* (/ provider (Math/abs provider) val)))

(defn set-angle
  "Rotates virtual space along y axis if true at given rpm per framrate."
  [angle rpm framerate]
  (if (< @angle rotation)
    (swap! angle + (/ (/ (* rpm rotation) 15) framerate))
    (swap! angle zero)))

(defn screenH [x y z ax ay az h0]
  (Math/round (+ (* x (Math/cos ax)) (* y (Math/cos ay)) (* z (Math/cos az)) h0)))

(defn screenV [x y z ax ay az v0]
  (Math/round (+ (* x (Math/sin ax)) (* y (Math/sin ay)) (* z (Math/sin az)) v0)))

(defn rotate [[x y] delta]
  (let [h (Math/hypot x y)
        a (Math/acos (/ x h))
        rot (+ a delta)]
    [(* (Math/cos rot) h) (* (Math/sin rot) h)]))

(defn point-cloud
  [a b c size mirror surface]
  "Calculate a set of 3D points per given surface function."

  ;; TODO: purify grid-size & grid-step
  (let [gx (range (- 0 size) size grid-x)
        gy (range (- 0 size) size grid-y)
        xs gx
        ys gy
        matrix (for [x xs y ys] [x y])
        ps (reduce into (map #(let [z (surface a b c %)]
                                (if (true? mirror)
                                  [[(first %) (second %) z] [(* -1 (first %)) (* -1 (second %)) z]
                                   [(first %) (second %) (* -1.0 z)] [(* -1 (first %)) (* -1 (second %)) (* -1.0 z)]] ; add mirrored points
                                  [[(first %) (second %) z] [(* -1 (first %)) (* -1 (second %)) z]])) (map #(rotate % @orient) matrix)))]
    ps))

(defn project
  "Project a collection of 3D points onto 2D screen."
  [points]
  (map #(let [[x y z] %]
          [(screenH x y z @Ax @Ay @Az originX) (screenV x y z @Ax @Ay @Az originY) (> y 0)]) points))

(defn setup []
  (q/frame-rate framerate)                    ;; Set framerate to 30 FPS
  (q/background 255 255 255))

(defn draw []
  (q/stroke-weight 1) (let [foreground (apply q/color (:fore-color config))
                            background (apply q/color (:aft-color config))
                            graph (project (point-cloud
                                            (nth (:constants config) 0)
                                            (nth (:constants config) 1)
                                            (nth (:constants config) 2)
                                            sheet-size
                                            (:mirror config)
                                            (:function config)))]
                        (q/clear)

                        ;; render axes
                        (if (true? axis?)
                          (do
                            (q/stroke (apply q/color lnf/snow-day))
                            (q/line (screenH 0 0 0 @Ax @Ay @Az originX) (screenV 0 0 0 @Ax @Ay @Az originY)
                                    (screenH xyz-length 0 0 @Ax @Ay @Az originX) (screenV xyz-length 0 0 @Ax @Ay @Az originY)) ; x-axis

                            (q/stroke (apply q/color lnf/umami))
                            (q/line (screenH 0 0 0 @Ax @Ay @Az originX) (screenV 0 0 0 @Ax @Ay @Az originY)
                                    (screenH 0 xyz-length 0 @Ax @Ay @Az originX) (screenV 0 xyz-length 0 @Ax @Ay @Az originY)) ; y-axis

                            (q/stroke (apply q/color lnf/fall-foliage))
                            (q/line (screenH 0 0 0 @Ax @Ay @Az originX) (screenV 0 0 0 @Ax @Ay @Az originY)
                                    (screenH 0 0 xyz-length @Ax @Ay @Az originX) (screenV 0 0 xyz-length @Ax @Ay @Az originY)))) ; z-axis

                        (q/set-pixel originX originY (q/color 0 255 0)) ;; Draw a point at the center of the screen

                        (doseq [[x y neg] graph]
                          (q/set-pixel x y (if (true? neg) background foreground))
                          (q/set-pixel (+ x 1) y (if (true? neg) background foreground))
                          (q/set-pixel (+ x 1) (+ y 1) (if (true? neg) background foreground))
                          (q/set-pixel x (+ y 1) (if (true? neg) background foreground)))

                        (if (true? render-frames)
                          (if (< @counter frame-count)
                            (do
                              (swap! counter inc)
                              (set-angle orient speed framerate)
                              (q/save (str "resources/cone-" @counter ".png"))))
                          (if (:animated config)
                            (set-angle orient speed framerate)))))


src/sparkl/surfaces.clj
;; ============================================================================
;; Quadric Surfaces
;; ============================================================================
(ns sparkl.surfaces)

(defn sqr [n]
  "Return square of the provided number."
  (* n n))

(defn paraboloid [a b c [x y]]
  "Reder a parabaloid."
  (/ (+ (/ (sqr x) (sqr a)) (/ (sqr y) (sqr b))) c))

(defn saddle [a b c [x y]]
  "Define a hyperbolic parabaloid."
  (/ (- (/ (sqr x) (sqr b)) (/ (sqr y) (sqr a))) c))

(defn cone [a b c [x y]]
  "Define a cone."
  (Math/sqrt (* (sqr c) (+ (/ (sqr x) (sqr a)) (/ (sqr y) (sqr b))))))

(defn one-sheet [a b c [x y]]
  "Define a hyperboloid of one sheet."
  (Math/sqrt (* (sqr c) (- (+ (/ (sqr x) (sqr a)) (/ (sqr y) (sqr b))) 1.0))))

(defn two-sheet [a b c [x y]]
  "Define a hyperboloid of two sheets."
  (Math/sqrt (* (sqr c) (- (/ (sqr x) (sqr a)) (/ (sqr y) (sqr b)) 1.0))))

(defn ellipsoid [a b c [x y]]
  "Define an ellipsoid."
  (Math/sqrt (* (sqr c) (- 1.0 (+ (/ (sqr x) (sqr a)) (/ (sqr y) (sqr b)))))))


src/sparkl/core.clj
(ns sparkl.core
  (:require [quil.core :as q :include-macros true]
            [sparkl.surfaces :as s]
            [clojure.pprint :as prnt]))

(defn zero [& args]
  0)

(def framerate 30)
(def speed 0.1) ; rpm

; (def radius (atom 1))
; (def inc-radius (atom true))
; (defn toggle [b]
;   (not b))
; (defn set-radius []
;   (if (> @radius 100)
;     (swap! inc-radius toggle)
;     (if (< @radius 1)
;       (swap! inc-radius toggle)))
;
;   (if (true? @inc-radius)
;     (swap! radius inc)
;     (swap! radius dec))
;   @radius)

(def axis? false)
(def originX (/ (q/screen-width) 2))
(def originY (/ (q/screen-height) 1.2))
(def xyz-length 128)
(def sheet-size 300)

;; Angles
(def Ax (atom (Math/toRadians 15)))
(def Ay (atom (Math/toRadians -15)))
(def Az (atom (Math/toRadians 270)))
(def orient (atom (Math/toRadians 0)))

(defn copy-sign [val provider]
  "Return val with sign of provider"
  (* (/ provider (Math/abs provider) val)))

(defn set-angle
  "Rotates virtual space along y axis if true at given rpm per framrate."
  [angle rpm framerate]
  (if (< @angle 400.0)
    (swap! angle + (/ (/ (* rpm 400.0) 60) framerate))
    (swap! angle zero)))

(defn screenH [x y z ax ay az h0]
  (Math/round (+ (* x (Math/cos ax)) (* y (Math/cos ay)) (* z (Math/cos az)) h0)))

(defn screenV [x y z ax ay az v0]
  (Math/round (+ (* x (Math/sin ax)) (* y (Math/sin ay)) (* z (Math/sin az)) v0)))

(defn rotate [[x y] delta]
  (let [h (Math/hypot x y)
        a (Math/acos (/ x h))
        rot (+ a delta)]
    [(* (Math/cos rot) h) (* (Math/sin rot) h)]))

(defn point-cloud
  [a b c size mirror surface]
  (let [step (range (- 0 size) size 20)
        cont (range (- 0 size) size 2)
        xs cont
        ys step
        matrix (for [x xs y ys] [x y])
        ps (reduce into (map #(let [z (surface a b c %)]
                                (if (true? mirror)
                                  [[(first %) (second %) z] [(* -1 (first %)) (* -1 (second %)) z]
                                   [(first %) (second %) (* -1.0 z)] [(* -1 (first %)) (* -1 (second %)) (* -1.0 z)]] ; add mirrored points
                                  [[(first %) (second %) z] [(* -1 (first %)) (* -1 (second %)) z]])) (map #(rotate % @orient) matrix)))]
    ps))

(defn circle
  [x y z r]
  ;; draw a circle
  ;; TODO: parametize z, right now z is 0
  (let [rs (range 0 (+ r 1) 20)
        xs (reduce into (map #(let [delta (Math/round (Math/sqrt (- (Math/pow r 2) (Math/pow % 2))))]
                                [[(+ x %) (+ y delta) 0]
                                 [(+ x %) (- y delta) 0]
                                 [(- x %) (+ y delta) 0]
                                 [(- x %) (- y delta) 0]]) rs))
        ys (reduce into (map #(let [delta (Math/round (Math/sqrt (- (Math/pow r 2) (Math/pow % 2))))]
                                [[(+ x delta) (+ y %) 0]
                                 [(+ x delta) (- y %) 0]
                                 [(- x delta) (+ y %) 0]
                                 [(- x delta) (- y %) 0]]) rs))]
    (set (reduce into [xs ys]))))

(defn project
  "Project a collection of 3D points onto 2D screen."
  [points]
  (map #(let [[x y z] %]
          [(screenH x y z @Ax @Ay @Az originX) (screenV x y z @Ax @Ay @Az originY) (> y 0)]) points))

(defn setup []
  (q/frame-rate framerate)                    ;; Set framerate to 30 FPS
  (q/background 0 0 0))

(defn draw []
  (q/stroke-weight 1) (let [white (q/color 255 255 255)
                            black (q/color 0 0 0)

                            stardust (q/color 63 137 155)
                            persimmon (q/color 188 62 49)

                            grape (q/color 193 155 179)
                            potato-chip (q/color 195 185 163)

                            ; good
                            wasabi (q/color 150 164 127)
                            nebula (q/color 75 70 67)

                            ; good
                            snow-day (q/color 204 204 204)
                            umami (q/color 102 102 102)

                            fall-foliage (q/color 198 205 194)
                            sriracha (q/color 189 14 26)

                            ; good
                            candlelight (q/color 206 174 71)
                            laughter (q/color 172 87 134)

                            foreground wasabi
                            background nebula

                            ; sphere (point-cloud 300.0 300.0 300.0 sheet-size true s/ellipsoid) ; screen-height / 2

                            graph (project (point-cloud 15.0 15.0 1 sheet-size false s/paraboloid))] ; screen-height / 1.2
                            ; graph (project (point-cloud 6 6 9 sheet-size false s/saddle))]
                            ; graph (project (point-cloud 6 6 9 sheet-size true s/cone))] ; screen-height / 2
                            ; graph (project (point-cloud 6 6 10 sheet-size true s/one-sheet))] ; screen-height / 2
                            ; graph (project (point-cloud 6 6 9 sheet-size true s/two-sheet))]
                            ; graph (project sphere)]
                        (q/clear)
    ;; render axis
                        (if (true? axis?)
                          (do
                            (q/stroke snow-day)
                            (q/line (screenH 0 0 0 @Ax @Ay @Az originX) (screenV 0 0 0 @Ax @Ay @Az originY)
                                    (screenH xyz-length 0 0 @Ax @Ay @Az originX) (screenV xyz-length 0 0 @Ax @Ay @Az originY)) ; x-axis

                            (q/stroke umami)
                            (q/line (screenH 0 0 0 @Ax @Ay @Az originX) (screenV 0 0 0 @Ax @Ay @Az originY)
                                    (screenH 0 xyz-length 0 @Ax @Ay @Az originX) (screenV 0 xyz-length 0 @Ax @Ay @Az originY)) ; y-axis

                            (q/stroke fall-foliage)
                            (q/line (screenH 0 0 0 @Ax @Ay @Az originX) (screenV 0 0 0 @Ax @Ay @Az originY)
                                    (screenH 0 0 xyz-length @Ax @Ay @Az originX) (screenV 0 0 xyz-length @Ax @Ay @Az originY)))) ; z-axis

                        (q/set-pixel originX originY (q/color 0 255 0)) ;; Draw a point at the center of the screen
                        (set-angle orient speed framerate)

                        (time (doseq [[x y neg] graph]
                                (q/set-pixel x y (if (true? neg) background foreground))
                                (q/set-pixel (+ x 1) y (if (true? neg) background foreground))
                                (q/set-pixel (+ x 1) (+ y 1) (if (true? neg) background foreground))
                                (q/set-pixel x (+ y 1) (if (true? neg) background foreground))))))

(q/defsketch quadric                 ;; Define a new sketch named example
             :title "Quadric Surfaces"    ;; Set the title of the sketch
             :settings #(q/smooth 2)             ;; Turn on anti-aliasing
             :setup setup                        ;; Specify the setup fn
             :draw draw                          ;; Specify the draw fn
             :features [:present]
             :size :fullscreen)

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  ; (prnt/pprint (map #(rotate % @orient) (point-cloud 10 10 1 sheet-size false s/paraboloid)))
  ; (prnt/pprint (point-cloud 10 10 1 sheet-size false s/paraboloid))
  (println "Rendering..."))
